fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -2717,6 +2722,346 @@ Future<Optional<std::string>> DataDistributionImpl::commit(ReadYourWritesTransac
+	state KeyRange range = kr.removePrefix(taskPrefixKey);
+	if (!normalKeys.contains(range)) {
+		TraceEvent(SevWarn, "BulkLoadCheckTaskStatusError")
+		    .detail("Reason", "Input range to check is out of normal range")
+		    .detail("InputRange", range);
+		throw bulkload_check_status_input_error();
+	}
+	// Check if there are existing updates in the current transaction
+	if (existingBulkLoadUpdate(ryw, Standalone(KeyRangeRef("task/"_sr, "task0"_sr))) ||
+	    existingBulkLoadUpdate(ryw, Standalone(KeyRangeRef("cancel/"_sr, "cancel0"_sr)))) {
+		TraceEvent(SevWarn, "BulkLoadCheckTaskStatusError")
+		    .detail("Reason", "Exist bulk loading update in the same transaction")
+		    .detail("InputRange", range);
+		throw bulkload_check_status_input_error();
+	}
+	RangeResult result = wait(krmGetRanges(&(ryw->getTransaction()), bulkLoadPrefix, range));
+	RangeResult res;
+	res.more = result.more;
+	res.readThrough = result.readThrough;
+	res.readToBegin = result.readToBegin;
+	res.readThroughEnd = result.readThroughEnd;
+	for (int i = 0; i < result.size(); ++i) {
+		Key keyToCopy = result[i].key.withPrefix(taskPrefixKey);
+		res.push_back_deep(res.arena(), KeyValueRef(keyToCopy, result[i].value));
+	}
+	return rywGetRange(ryw, kr, res);
+}
+
+Future<RangeResult> BulkLoadStatusImpl::getRange(ReadYourWritesTransaction* ryw,
+                                                 KeyRangeRef kr,
+                                                 GetRangeLimits limitsHint) const {
+	return BulkLoadStatusGetRangeActor(ryw, getKeyRange().begin, kr);
+}
+
+BulkLoadTaskImpl::BulkLoadTaskImpl(KeyRangeRef kr) : SpecialKeyRangeRWImpl(kr) {}
+
+Future<RangeResult> BulkLoadTaskImpl::getRange(ReadYourWritesTransaction* ryw,
+                                               KeyRangeRef kr,
+                                               GetRangeLimits limitsHint) const {
+	throw not_implemented();
+}
+
+void BulkLoadTaskImpl::clear(ReadYourWritesTransaction* ryw, const KeyRangeRef& range) {
+	throw not_implemented();
+}
+
+void BulkLoadTaskImpl::clear(ReadYourWritesTransaction* ryw, const KeyRef& key) {
+	throw not_implemented();
+}
+
+void BulkLoadTaskImpl::set(ReadYourWritesTransaction* ryw, const KeyRef& key, const ValueRef& value) {
+	Key taskPrefixKey =
+	    SpecialKeySpace::getModuleRange(SpecialKeySpace::MODULE::BULKLOADING).begin.withSuffix("task/"_sr);
+	BulkLoadState bulkLoadState = decodeBulkLoadState(value);
+	if (key != taskPrefixKey) {
+		TraceEvent(SevWarn, "BulkLoadSetTaskError")
+		    .detail("Reason", "Input key error")
+		    .detail("CorrectKey", taskPrefixKey)
+		    .detail("InputKey", key)
+		    .detail("InputState", bulkLoadState.toString());
+		throw bulkload_add_task_input_error();
+	}
+	ASSERT(bulkLoadState.isValid());
+	KeyRangeRef bulkLoadRange = bulkLoadState.range;
+	ASSERT(!bulkLoadRange.empty());
+	if (bulkLoadRange.begin >= normalKeys.end || bulkLoadRange.end >= normalKeys.end) {
```

Comment:
can be simplified as

```suggestion
	if (bulkLoadRange.end >= normalKeys.end) {
```